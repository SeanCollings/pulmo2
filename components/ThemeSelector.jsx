import React, { useContext } from 'react';
import { StyleSheet, View, BackHandler } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import { THEME, SettingsContext } from '../context/settings-context';
import CustomSwitch from './Switch';
import { DARK_MODE, LIGHT_MODE } from '../constants/constants';
import { useTheme } from '../hooks/useTheme';

const ThemeSelector = (props) => {
  const theme = useTheme();
  const { settingsContext, updateSettingsContext } = useContext(
    SettingsContext
  );

  const selectedTheme = settingsContext[THEME];

  const onChangeHandler = (active) => {
    const selectedTheme = active ? DARK_MODE : LIGHT_MODE;
    updateSettingsContext(THEME, selectedTheme);
  };

  return (
    <View style={styles.container}>
      <View style={styles.iconContainer}>
        <MaterialCommunityIcons
          name={'theme-light-dark'}
          color={theme.BORDER}
          size={22}
        />
      </View>
      <CustomSwitch
        onChange={onChangeHandler}
        state={selectedTheme === DARK_MODE}
        trackColor={theme.SECONDARY}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
  },
  iconContainer: {
    justifyContent: 'center',
  },
});

export default ThemeSelector;
