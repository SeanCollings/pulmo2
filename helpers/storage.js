import { AsyncStorage } from 'react-native';
import { APP_ID } from '../constants/constants';

export const clearAllAsyncStorage = () => {
  AsyncStorage.clear();
};

export const getAsyncData = async (id) => {
  try {
    const value = await AsyncStorage.getItem(`${APP_ID}${id}`);
    if (value !== null) {
      return JSON.parse(value);
    } else {
      return null;
    }
  } catch (error) {
    console.log('Failure retrieving ' + id + ' error: ' + error);
  }
};

export const getMultiAsyncData = async (id, keys) => {
  try {
    const formattedKeys = keys.map((key) => `${APP_ID}${id}${key}`);
    const values = await AsyncStorage.multiGet(formattedKeys);

    if (values && !!values.length) return values;
    else {
      return null;
    }
  } catch (err) {
    console.log('Failure retrieving from all keys with error ' + error);
  }
};

export const storeAsyncData = async (id, value) => {
  try {
    await AsyncStorage.setItem(`${APP_ID}${id}`, JSON.stringify(value));
    return true;
  } catch (error) {
    console.log('Failure saving: ' + error);
    return false;
  }
};

export const removeAsyncData = async (id) => {
  try {
    await AsyncStorage.removeItem(`${APP_ID}${id}`);
  } catch (err) {
    console.log('Failure removing: ' + error);
  }
};
