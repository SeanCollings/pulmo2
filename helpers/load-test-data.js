import testData, {
  TEST_KEY,
  TEST_TITLE,
  TEST_MUTLIPLE_ID,
  TEST_DATA,
  TEST_PRE_ID,
} from '../data/test-data';
import { getAsyncData, storeAsyncData } from './storage';

export const MESSAGE_UNKNOWN = 'unknown';
export const MESSAGE_PRE_SAVED = 'pre-saved';
export const MESSAGE_DONE = 'done';

export default async (title) => {
  try {
    const selectedData = Object.values(testData).find(
      (data) => data[TEST_TITLE] === title
    );

    if (!selectedData) return MESSAGE_UNKNOWN;

    if (!selectedData[TEST_MUTLIPLE_ID]) {
      const savedData = await getAsyncData(selectedData[TEST_KEY]);

      if (!savedData) {
        console.log('Saving data for:', selectedData[TEST_KEY]);
        await storeAsyncData(selectedData[TEST_KEY], selectedData[TEST_DATA]);
        return MESSAGE_DONE;
      } else {
        return MESSAGE_PRE_SAVED;
      }
    } else {
      const updatedSaves = await selectedData[TEST_DATA].reduce(
        async (acc, data) => {
          const newAcc = await acc;
          const id = selectedData[TEST_MUTLIPLE_ID];
          const preId = selectedData[TEST_PRE_ID] || '';
          const savedData = await getAsyncData(`${preId}${data[id]}`);

          if (!savedData) {
            console.log('Saving data for:', `${preId}${data[id]}`);
            newAcc.push(data[id]);
            await storeAsyncData(`${preId}${data[id]}`, data);
          }
          return newAcc;
        },
        []
      );

      if (!updatedSaves.length) return MESSAGE_PRE_SAVED;

      return `${MESSAGE_DONE} [${updatedSaves.length}]`;
    }
  } catch (err) {
    throw new Error(err);
  }
};
