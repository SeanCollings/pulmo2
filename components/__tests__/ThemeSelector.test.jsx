import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup, fireEvent } from 'react-native-testing-library';
import ThemeSelector from '../ThemeSelector';
import SettingsContextProvider from '../../context/settings-context';
import { storeAsyncData } from '../../helpers/storage';

jest.mock('../../helpers/storage');

describe('Slides - unit test', () => {
  // Hide console.error that I can't find a reason for yet:
  // Warning: dispatchCommand was called with a ref that isn't a native component.
  // Use React.forwardRef to get access to the underlying native component
  jest.spyOn(console, 'error').mockImplementation(jest.fn());

  let component;

  afterEach(() => {
    cleanup();
  });

  test('should render', async () => {
    await act(async () => {
      component = create(<ThemeSelector />);
    });
    expect(component).toMatchSnapshot();
  });

  test('should enable theme change', async () => {
    storeAsyncData.mockReturnValueOnce(Promise.resolve());

    await act(async () => {
      component = create(
        <SettingsContextProvider>
          <ThemeSelector />
        </SettingsContextProvider>
      );
    });

    const valueChange = component.root.find(
      ({ props }) => props.testID === 'customSwitch_valueChange'
    );

    fireEvent(valueChange, 'onValueChange', true);
    expect(component).toMatchSnapshot();
  });
});
