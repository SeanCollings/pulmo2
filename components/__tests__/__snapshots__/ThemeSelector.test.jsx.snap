// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slides - unit test should enable theme change 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ededed",
        "fontFamily": "tit-regular",
        "fontSize": 17,
        "opacity": 0.87,
      }
    }
  >
    Use dark mode
  </Text>
  <View>
    <View
      style={
        Object {
          "paddingTop": 1,
        }
      }
    >
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="#2f9fd0"
        style={
          Object {
            "height": 31,
            "width": 51,
          }
        }
        testID="customSwitch_valueChange"
        thumbTintColor=""
        tintColor="#707070"
        value={true}
      />
    </View>
  </View>
</View>
`;

exports[`Slides - unit test should render 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#3e4b57",
        "fontFamily": "tit-regular",
        "fontSize": 17,
        "opacity": 1,
      }
    }
  >
    Use dark mode
  </Text>
  <View>
    <View
      style={
        Object {
          "paddingTop": 1,
        }
      }
    >
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="#002f56"
        style={
          Object {
            "height": 31,
            "width": 51,
          }
        }
        testID="customSwitch_valueChange"
        thumbTintColor=""
        tintColor="#CCC"
        value={false}
      />
    </View>
  </View>
</View>
`;
